### PR TITLE
Fixes to allow FV3_HRRR_c3 to run with gnu debug plus PR #113, #106, and #103

### LINKED ISSUE #101: FV3_HRRR_c3 suite crashes in c3 calls on Cheyenne with gnu compiler in debug mode.
# Description

The FV3_HRRR_c3 suite crashes in c3 calls on Cheyenne when compiled with the gnu compiler in debug mode.

## Steps to Reproduce

Please provide detailed steps for reproducing the issue.

1. Create a hrrr_c3_debug variant of hrrr_control_debug in the ufs weather model regression tests
2. Run it on Cheyenne with the gnu compiler.
3. Model crashes.

## Additional Context

Please provide any relevant information about your setup. This is important in case the issue is not reproducible except for under certain conditions.

* Machine - Cheyenne
* Compiler - gnu
* Suite Definition File or Scheme - FV3_HRRR_c3
* fixed by #102

### LINKED ISSUE #104: incorrect filename in warning message for missing noahmptable.tbl
# Description

When the model requires `noahmptable.tbl`, and it isn't present, the error message is wrong. The message asks for `noahmptable.tb` without the final `l`.

```f90
    inquire( file='noahmptable.tbl', exist=file_named )
    if ( file_named ) then
       open(15, file="noahmptable.tbl", status='old', form='formatted', action='read', iostat=ierr)
    else
       open(15, status='old', form='formatted', action='read', iostat=ierr)
    end if
    if ( ierr /= 0 ) then
       errmsg = 'warning: cannot find file noahmptable.tb' !<---- ERROR IS HERE
       errflg = 1
       return
!      write(*,'("warning: cannot find file noahmptable.tbl")')
    endif
```

## Steps to Reproduce

Please provide detailed steps for reproducing the issue.

1. Run a NOAHMP configuration without noahmptable.tbl
2. Get the wrong error message

## Additional Context

* Machine - N/A
* Compiler - N/A
* Suite Definition File or Scheme - anything that uses noahmp
* Reference other issues or PRs in other repositories that this is related to, and how they are related.

### LINKED ISSUE #105: GFS_phys_time_vary_init does not report errmsg/errflg correctly due to thread race condition
# Description

The GFS_phys_time_vary_init is parallelized using mpi sections, but it does not correctly handle errmsg or errflg. All threads update the same errmsg and errflg. That means a failure message can be overwritten by a success message in a later step.

Pinging @hu5970 who encountered the problem and @tanyasmirnova who is debugging it too.

To visualize this, suppose there are two threads running at once. For simplicity's sake, lets say there are only two initialization calls: init_that_fails() and init_that_succeeds()

#### Failure happens first

Events happened in this order:

Thread 1: Completes init_that_fails() and sets errmsg=1
Thread 2: Completes init_that_succeeds() and sets errmsg=0

The errmsg is 0 and the model will run even though one of the initialization steps failed.

#### Failure happens second

Events happened in this order:

Thread 2: Completes init_that_succeeds() and sets errmsg=0
Thread 1: Completes init_that_fails() and sets errmsg=1

The errmsg is 1 so the model will abort as expected.

## Steps to Reproduce

Please provide detailed steps for reproducing the issue.

1. Delete noahmptable.tbl
2. Use a scheme that does not require that file.
3. Run the model a few times with at least two threads.
4. Notice that it fails sporadically instead of 100% of the time.

## Additional Context

This was discovered in an RRFS parallel. The machine, compiler, etc. doesn't matter. However, the easiest way to see it is to run a non-NOAHMP suite without noahmptable.tbl.

### LINKED ISSUE #107: The `noahmptable.tbl` file is mandatory, even for configurations that don't need it.
# Description

The GFS_phys_time_vary.fv3.F90 reads the `noahmptable.tbl` file unconditionally. Even configurations that don't need that table will try to read it. The code tries to set errflg=1 if the file is missing. (Though, as of this writing, that file's error checking is broken: #105)

## Steps to Reproduce

Please provide detailed steps for reproducing the issue.

1. Run a non-noahmp suite many times.
2. Some of the runs will fail. It should be 100% of the runs, but the error checking is broken (#105) so errors in that routine are usually not reported.

## Additional Context

Discovered by @hu5970 in the RRFS parallels

### LINKED ISSUE #108: ozinterp and h2ointerp data is read when it is not needed
# Description

The ozinterp and h2ointerp data is read unconditionally in GFS_time_vary_pre.fv3.F90. This makes the files mandatory, even for configurations that don't need them. That error may not be caught due to bug #105 but it will be caught after that bug is fixed.

## Steps to Reproduce

Please provide detailed steps for reproducing the issue.

1. Choose a suite that doesn't need one of those files, and ensure the file is missing.
2. Run it many times. It'll succeed frequently due to bug #105 but it'll eventually fail.

### LINKED ISSUE #112: CLM Lake restarts nine 3D fields that are constant
# Description

These fields in the CLM Lake model are constant:

- lake_z3d
- lake_dz3d
- lake_soil_watsat3d
- lake_csol3d
- lake_soil_tkmg3d
- lake_soil_tkdry3d
- lake_soil_tksatu3d
- lake_clay3d
- lake_sand3d

The later seven are 2D arrays that are needlessly duplicated in the vertical (lake k) dimension. The first two are 3D. All of these are sparse arrays, zero everywhere except lake points.

All nine should be removed.

## Steps to Reproduce

1. Read the code. The lakeini in clm_lake.f90 calculates those variables. They're never modified after that.

## Additional Context

The CLM Lake Model is taking up too much disk space in the RRFS. Removing these nine variables would reduce that problem.

Problem identified by @hu5970 and @tanyasmirnova

### PR DESCRIPTION
**Pull request by SamuelTrahanNOAA**
The c3 code cannot handle null pointers correctly. This changes some array argument declarations so tests pass. At the ufs-weather-model level, I've added hrrr_c3_debug tests for both the gnu and intel compilers. The new tests pass on Jet, Hera, and Cheyenne.

This PR includes #113 and #106 

- Fixes #112
- Fixes #101
- Fixes #108
- Fixes #107
- Fixes #105
- Fixes #104 
- Adds `errflg=0` and `errmsg=''` in two routines that were missing them.

### CODE REVIEW

**Comment by lisa-bengtsson** [comment]
Changes look good, but it is a bit surprising the the cu_gf_deep.F90 doesn't run into the same problem?

**Comment by SamuelTrahanNOAA** [comment]
> Changes look good, but it is a bit surprising the the cu_gf_deep.F90 doesn't run into the same problem?

Correct. I didn't change other schemes because I wanted to simplify the changes as much as possible.

My ufs-weather-model changes add hrrr_c3_debug and hrrr_gf_debug tests for both compilers. If we encounter problems again, we'll know.

**Review by grantfirl — APPROVED** [review]

**Review by haiqinli — APPROVED** [review]

**Review by lisa-bengtsson — APPROVED** [review]

**Comment by SamuelTrahanNOAA** [comment]
In a recent email about this PR, I got the usual question about WRF compatibility, so I'm pasting my answer here.

WRF needs a starting index for each dimension. By that, I mean the `its:` part of `its:ite`. If you remove both the `its` and the `ite` (`dimension(:)`) then the starting index is 1. That breaks WRF. However, if you keep the starting index (`dimension(its:)`) then the array starts at `its` and WRF will work. If you include the `ite` then CCPP will crash with the gnu compiler if you pass a null pointer.

In short, your dimensions must look like this

```
dimension(its:,kts:) :: my_3d_array
dimension(its:) :: my_2d_array
dimension(kts:) :: my_column_array
```

That way, both WRF and CCPP work.

**Review by Qingfu-Liu — APPROVED** [review]

**Comment by dustinswales** [comment]
> In a recent email about this PR, I got the usual question about WRF compatibility, so I'm pasting my answer here.
> 
> WRF needs a starting index for each dimension. By that, I mean the `its:` part of `its:ite`. If you remove both the `its` and the `ite` (`dimension(:)`) then the starting index is 1. That breaks WRF. However, if you keep the starting index (`dimension(its:)`) then the array starts at `its` and WRF will work. If you include the `ite` then CCPP will crash with the gnu compiler if you pass a null pointer.
> 
> In short, your dimensions must look like this
> 
> ```
> dimension(its:,kts:) :: my_3d_array
> dimension(its:) :: my_2d_array
> dimension(kts:) :: my_column_array
> ```
> 
> That way, both WRF and CCPP work.

@SamuelTrahanNOAA Your are spot on.

**Review by dustinswales — APPROVED** [review]

**Comment by dudhia** [comment]
@dustinswales does it actually break WRF? We just like to keep the absolute index for debugging purposes.

**Comment by dustinswales** [comment]
> e absolute index for debugging purposes.

@dudhia 
@SamuelTrahanNOAA is the expert on null pointers :)
But from my understanding, 
changing from its:ite -> : will break the WRF code, but the CCPP will work
changing from its:ite -> its: will work for both hosts.

**Comment by SamuelTrahanNOAA** [comment]
> does it actually break WRF? We just like to keep the absolute index for debugging purposes.

Loops inside the WRF physics use its as a starting index. If the arrays start at 1 and you access index its, it may be far beyond the allocated array. The its:ite may be 2560:2610. Lose the its and the array bounds will be 1:51. Accessing index its=2560 puts you past the array bounds.

Also, WRF arrays are declared with memory bounds, not tile bounds:

`dimension(ims:ime,kms:kme,jms:jme)`

So long WRF passes arrays to the low-level physics routines like this, it'll work:

`call thing(grid%var(its:ite,kts:kte,jts:jte))`
or
`call thing(var(its:ite,kts:kte,jts:jte))`

**Comment by dudhia** [comment]
@SamuelTrahanNOAA I see. Yes, we can't do that in WRF as we also pass in _its_. CCPP must be treating those as 1.

**Comment by SamuelTrahanNOAA** [comment]
I've merged #106 into this PR since the UFS code managers plan to merge the two simultaneously.

**Comment by BrianCurtis-NOAA** [comment]
Can we get at least one re-approval since there have been code changes since the latest approval. Just need someone to be OK with those newer changes.

**Review by grantfirl — APPROVED** [review]
@BrianCurtis-NOAA This still looks good to me after the combination of PRs.

**Review by grantfirl — APPROVED** [review]

**Review by grantfirl — APPROVED** [review]

**Comment by SamuelTrahanNOAA** [comment]
@haiqinli - Please confirm I merged your changes correctly.

**Comment by haiqinli** [comment]
@SamuelTrahanNOAA Thank you very much! Your merging looks good to me.

**Comment by jkbk2004** [comment]
@grantfirl @dustinswales All tests are done at https://github.com/ufs-community/ufs-weather-model/pull/1893. Can you merge this pr?
